py/objlist.c

```diff
@@ -156,7 +156,7 @@ STATIC mp_obj_t list_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value) {
         // delete
 #if MICROPY_PY_BUILTINS_SLICE
         if (MP_OBJ_IS_TYPE(index, &mp_type_slice)) {
-            mp_obj_list_t *self = self_in;
+            mp_obj_list_t *self = MP_OBJ_CAST(self_in);
```

Comment:
If MP_OBJ_CAST is something like (self_in+4) and self_in is never referenced again in this function, then the compiler may optimise it such that self reuses the stack/register that self_in previously used.  Thus, no pointer to the original self_in exists any more and the object may be GC'd.